test/common/network/udp_listener_impl_batch_writer_test.cc

```diff
@@ -0,0 +1,300 @@
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+/**
+ * Tests UDP Packet Writer To Send packets in Batches to a client
+ *  1. Setup a udp listener and client socket
+ *  2. Send different sized payloads to client.
+ *     - Verify that the packets are buffered as long as payload
+ *       length matches gso_size.
+ *     - When payload size > gso_size verify that the new payload is
+ *       buffered and already buffered packets are sent to client
+ *     - When payload size < gso_size verify that the new payload is
+ *       sent along with the already buffered payloads.
+ *  3. Call UdpPacketWriter's External Flush
+ *     - Verify that the internal buffer is emptied and its contents
+ *       are delivered to the client.
+ */
+TEST_P(UdpListenerImplBatchWriterTest, SendData) {
+  EXPECT_TRUE(listener_->udpPacketWriter()->isBatchMode());
+  Address::InstanceConstSharedPtr send_from_addr = getUnlikelySourceAddress();
+
+  absl::FixedArray<std::string> payloads{"length7", "length7", "len<7",
+                                         "length7", "length7", "length>7"};
+  std::string internal_buffer("");
+  std::string first_buffered("");
+  std::list<std::string> pkts_to_send;
+
+  // Get initial value of total_bytes_sent
+  uint64_t total_bytes_sent =
+      listener_config_.listenerScope().counterFromString("total_bytes_sent").value();
+
+  for (const auto& payload : payloads) {
+    Buffer::InstancePtr buffer(new Buffer::OwnedImpl());
+    buffer->add(payload);
+    UdpSendData send_data{send_from_addr->ip(), *client_.localAddress(), *buffer};
+
+    auto send_result = listener_->send(send_data);
+    EXPECT_TRUE(send_result.ok()) << "send() failed : " << send_result.err_->getErrorDetails();
+
+    // Verify udp_packet_writer stats for batch writing
+    if (internal_buffer.length() == 0 ||        /* internal buffer is empty*/
+        payload.compare(first_buffered) == 0) { /*len(payload) == gso_size*/
+      pkts_to_send.emplace_back(payload);
+      internal_buffer.append(payload);
+      first_buffered = payload;
+    } else if (payload.compare(first_buffered) < 0) { /*len(payload) < gso_size*/
+      pkts_to_send.emplace_back(payload);
+      internal_buffer.clear();
+      first_buffered.clear();
+    } else { /*len(payload) > gso_size*/
+      internal_buffer = payload;
+      first_buffered = payload;
+    }
+
+    EXPECT_EQ(listener_config_.listenerScope()
+                  .gaugeFromString("internal_buffer_size", Stats::Gauge::ImportMode::NeverImport)
+                  .value(),
+              internal_buffer.length());
+    EXPECT_EQ(listener_config_.listenerScope()
+                  .gaugeFromString("front_buffered_pkt_size", Stats::Gauge::ImportMode::NeverImport)
+                  .value(),
+              first_buffered.size());
+
+    if (send_result.rc_ > 0) {
+      // Verify Correct content is received at the client
```

Comment:
This doesn't really test gso behavior because even if w/o GSO, client_.recv() would still read all the packets previously sent by the listener.

Reply:
The idea behind having this here is to verify that the total_packets_sent parameter is only updated when packets are actually sent. Modified this part to be more concise at checking this now!

Reply:
Added a new comment for this.
```
    // Verify that the total_bytes_sent is only updated when the packets
    // are actually sent to the client, and not on being buffered.
```